### ISSUE TITLE
Hibernate's `@Version` annotation to enable Optimistic Locking is ignored

### ISSUE THREAD

**Issue by manoelcampos** [quarkusio/quarkus#7193]
**Describe the bug**
The Hibernate´s `@Version` annotation is having no effect when using Panache.

**Expected behavior**
An `OptimisticLockException` should be thrown to avoid possible data override when two users try to update the same entity at the same time. 

**Actual behavior**
When using Hibernate's `@Version` annotation to enable Optimistic Locking, if two users change the same Entity at the same time, no `OptimisticLockException` is thrown.

**To Reproduce**

To make it easy to reproduce the issue, a repository was created [here](https://github.com/manoelcampos/hibernate-concurrency). It provides two minimal applications to show the issue. The applications use an in-memory H2 database, requiring no extra configurations.

**Environment (please complete the following information):**
 - Quarkus version: 1.2.0.Final

**Comment by emmanuelbernard** [comment]
Can you try one thing @manoelcampos 

Add 
https://github.com/manoelcampos/hibernate-concurrency/blob/master/2-database-concurrency-panache/src/main/java/com/manoelcampos/server/model/Cliente.java#L23

`existente.version = cliente.version;`

to make sure we get the client side version copied so we detect the conflict.

Also replace https://github.com/manoelcampos/hibernate-concurrency/blob/master/2-database-concurrency-panache/src/main/java/com/manoelcampos/server/model/Cliente.java#L27 `existente.persist()` with `existente.flush()` and see what's happening.

You must do flush() so that the operation is effectively sent to the database. persist on an existing entity is a noop.

In practice, your v1 is doing a merge of a client side copy and thus the copy is automatic.

BTW @FroMage , there is no merge equivalent so the copy is manual. Should we have a reattach operation of sort?

Also, you should have a `@Transactional` here https://github.com/manoelcampos/hibernate-concurrency/blob/master/2-database-concurrency-panache/src/main/java/com/manoelcampos/server/rest/ClienteResource.java#L37 to ensure that a flush and an actual update of the database happen. But that change won't fix your optimistic locking.

**Comment by manoelcampos** [comment]
Hello @emmanuelbernard 
In fact, I removed the field from the update just to try different things and forgot to put it back. Including the field is the way it's supposed to work, but it doesn't.
I just updated the source code including it.

I added 2 animated gifs to the [README](https://github.com/manoelcampos/hibernate-concurrency) to show the issue.

Replacing `persist()` by `flush()` or `persistAndFlush()` doesn't have any effect with the locking. The `@Version long versao` field is incremented, but the data from the previous user is overwritten.

About the transaction, as it's just an example, I made the entire `ClienteResource` class transactional. I moved the `@Transactional` directly to the `update` method and it doesn't help anyway as you said.

I made sure to stop and restart Quarkus after such changes.

**Comment by emmanuelbernard** [comment]
@Sanne @gsmet I am a bit lost but it seems that

```
Foo fooWithChangesAndOldVersion = ...;
Foo foo = em.find(Foo.class, 1L);
foo.title = fooWithChangesAndOldVersion.title;
foo.version = fooWithChangesAndOldVersion.version;
em.flush();
```
Does not raise an exception because the version number overridden is ignored in the update SQL comparison. The comparison is there, just the version value loaded into foo is used which defeats the optimistic versioning mechanism.
I have to admit I am blanking whether that's expected behavior or not. In JPA we don't have that problem as merge does an explicit checking. 

Thoughts? Is that expected Hibernate behavior, a hibernate bug? Or some unexpected interaction between Panache and Hibernate?

**Comment by manoelcampos** [comment]
@emmanuelbernard 
I've just found the issue.
Panache doesn't expose the `EntityManager.update()` method, so we have to write extra code to merge an existing entity:

```java
public static boolean update(Cliente cliente) {
        Cliente existing = Cliente.findById(cliente.id);
        if(existing != null){
            existing.nome = cliente.nome;
            existing.cpf = cliente.cpf;
            existing.sexo = cliente.sexo;
            existing.endereco = cliente.endereco;
            existing.telefone = cliente.telefone;
            existing.versao = cliente.versao;
            existing.persistAndFlush();
            return true;
        }

        return false;
    }
```

I just added a `merge()` method to `JpaOperations` and `PanacheEntity` classes and it works as expected. This way, I can update an entity by just calling: 

```java
    public static boolean update(Cliente cliente) {
        cliente.merge();
        cliente.flush();
        return true;
    }
```

I don't know why the `EntityManager.merge()` isn't exposed in Panache.
I just added it in my fork at https://github.com/manoelcampos/quarkus/tree/hibernate_optimistic_locking
I can send a PR if you think it's OK.

**Comment by emmanuelbernard** [comment]
I don't think it's ok no, at least I'm not sure. I need to discuss with @FroMage and @Sanne on this one.

**Comment by FroMage** [comment]
Given:

```java
Foo fooWithChangesAndOldVersion = ...;
Foo foo = em.find(Foo.class, 1L);
foo.title = fooWithChangesAndOldVersion.title;
foo.version = fooWithChangesAndOldVersion.version;
em.flush();
```

It appears that `foo.version = fooWithChangesAndOldVersion.version` differs in behaviour from what the equivalent `merge` would do.

My guess is that setting the version manually causes some dirty state to be set while `merge` setting the version does not. We have to look at the code to make sure.

As to why we didn't add `merge` is that we figured that detached entities are evil and the source of too many bugs and WTFs. Now, admittedly this issue is a WTF that would be fixed by adding `merge` but we did not expect that ;)

**Comment by Sanne** [comment]
> It appears that foo.version = fooWithChangesAndOldVersion.version differs in behaviour from what the equivalent merge would do.

That's spot on: the Hibernate ORM documentation specifically prohibits messing with the version attribute - it's going to be ignored. I wonder if it shouldn't throw an exception.

I suspect you'll need to expose the _merge()_ operation; but I agree that it would be nice to find an alternative.

**Comment by FroMage** [comment]
I can probably make it throw if you attempt to set the version to a different version if the entity is managed.

Is it `merge` that throws, or is it on `flush`?

**Comment by manoelcampos** [comment]
@FroMage yeah, it's really strange. If I can help anyway, just let me know. I'll be glad to.
@Sanne, to avoid exposing  `merge()`, the `persist()` method could be changed as below, but that will change the behavior of `persist()` and may cause unexpected surprises for existing users:

```java
public static void persist(Object entity) {
    EntityManager em = getEntityManager();
    if(isPersistent(entity))
          merge(entity);
    else persist(em, entity);
}
```

**Comment by Sanne** [comment]
If persist just behaves like merge, I'd prefer it to be called merge to not confuse people.

BTW are you sure that would work? I expect `isPersistent` would always return false for a detached entity.

**Comment by Sanne** [comment]
> I can probably make it throw if you attempt to set the version to a different version if the entity is managed.
> 
> Is it `merge` that throws, or is it on `flush`?

I guess ideally I'd throw directly on attempting to write on the field.

But while adding such extra checks might avoid some confusion, it doesn't solve the problem

**Comment by FroMage** [comment]
> I guess ideally I'd throw directly on attempting to write on the field.

That should be easy

> But while adding such extra checks might avoid some confusion, it doesn't solve the main problem: expose a way to attache detached entities, or alternatively give people an explicit way to deserialize an entity including its version state.

But this is not the problem here. This is an orthogonal issue which we are investigating as part of the resteasy/panache problem space in another issue.

**Comment by manoelcampos** [comment]
> BTW are you sure that would work? I expect isPersistent would always return false for a detached entity.

@Sanne, in fact, `isPersistent()` will return false for detached entities. What I usually do in my applications is just:

```java
public static void persist(Object entity) {
    EntityManager em = getEntityManager();
    if(entity.id != null and entity.id > 0)
          merge(entity);
    else persist(em, entity);
}
```

**Comment by manoelcampos** [comment]
> Is it merge that throws, or is it on flush?

@FroMage, merge is the one to throw the `OptimisticLockException`, as can be seen in the stacktrace below:

```bash
javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [com.manoelcampos.server.model.Cliente#1]
        at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:223)
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:93)
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
        at org.hibernate.internal.SessionImpl.fireMerge(SessionImpl.java:786)
        at org.hibernate.internal.SessionImpl.merge(SessionImpl.java:762)
        at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.merge(TransactionScopedEntityManager.java:124)
        at io.quarkus.hibernate.orm.runtime.entitymanager.ForwardingEntityManager.merge(ForwardingEntityManager.java:32)
        at com.manoelcampos.server.dao.JpaDAO.save(JpaDAO.java:31)
```

**Comment by emmanuelbernard** [comment]
Let me take a step back.

`PanacheEntity` objects are where operations live.
JPA `merge()` by default, will return a _copy_ of a detached entity.
And the use case we want to fix is to be able to copy part or the entirety of the state from a version coming from the web back into the database.

Option A

So we can expose `panacheEntity.merge()` knowing that the merge semantic will not be the same as JPAs (no copy here). Also I wonder what `PanacheRepository` should do.

Option A1

We can expose a `panacheEntity.reattach()` to clarify that the semantic is different.

PS what do we do semantic wise if there is already an existing entity in the EM context and we want to "pseudo-merge" or reattach (option A or A1).

None do solve the problem of partial updates.

```
@Entity 
public class Foo extends PanacheEntity {
    @Version public long version;
    public String title;
    public String author;
}

@PUT
public updateTitle(Foo foo) {
    Foo existing = Foo.findById(foo.id);
    existing.title = foo.title;
    existing.version = foo.version;
    // not author!
    existing.merge(); // is that how we want to do it? 
}
```

This is a different semantic from Hibernate ORM (version protection) and I'd like to know if we are fine with all that.

**Comment by emmanuelbernard** [comment]
Option B is to not have full state merge and force a manual copy.

```
Foo existing = Foo.findById(foo.id);
existing.title = foo.title;
existing.version = foo.version;
existing.author = foo.author;
existing.merge(); // or update() maybe
// or have flush() check it and do some entity tracking above / or in Hibernate??
```

**Comment by Sanne** [comment]
> But this is not the problem here. This is an orthogonal issue which we are investigating as part of the resteasy/panache problem space in another issue.

@FroMage I'm not following - what is this `this` that you're referring to? Please quote when commenting. Your previous comment `I can probably make it throw if you attempt to set the version to a different version` seemed to refer to my suggestion that ORM should not really let people write on the version field, since it ignores it. I guess a "WARN: you've messed with field x, I'm going to ignore it" could also work.  Either way, "this is not the problem here" is exactly what I said: `it doesn't solve the main problem`

@manoelcampos :

> in fact, isPersistent() will return false for detached entities. What I usually do in my [..]

Yes I understand, but while that code might work for you, it isn't safe for other use cases; for example when using explicitly assigned identifiers. We need a bullet-proof solution to be able to rely it in framework code.

So let's not overload the meaning of `persist()`, this needs to be an explicit user choice.

@emmanuelbernard Option B is defeating the point of optimistic version control :) In Manoel's example, the webpage would present data from entity version 1, but then you load entity version 2 and copy changes to it. You need the "original" existing entity at version 1.

### Option A1

+1 Seems to be the most sensible choice.

We need to avoid the case of an existing attached instance with the same ID: maybe just throw an exception? I think I'd use the exception as I'd consider it an unexpected use case.

**Comment by Sanne** [comment]
hum actually throwing an exception might not be straight forward as there's grey areas when it comes to associations: what if some of the objects in the model already exist in the session, and others don't? Is it ok to fail only after having loaded some of these?

**Comment by emmanuelbernard** [comment]
To clarify, in option B we would have a different behavior of Hibernate (or of the panache layer): i.e. let the version be overridden and checked before.

Option C

To avoid the attach problem I think I would prefer the original merge contract 

```
class PanacheEntity
    // should be called mergeState?
    public static PanacheEntity merge(PanacheEntity panacheEntity) {
        return merge(panacheEntity);
    }
}
```

The reason is that Hibernate had reattach semantic in the past but we moved away from it. I remember @gavinking saying that the saveOrUpdate semantics were dodgy and that might have been the reasons.

For partial update, then I think the user should findById and do the manual version comparison. I don't know how we could nicely help otherwise (except maybe in RESTEasy with Panache

**Comment by manoelcampos** [comment]
> To avoid the attach problem I think I would prefer the original merge contract

@emmanuelbernard 

I'm thinking that it would be great if `merge()` was a generic method such as below:

```java
    public <T extends PanacheEntityBase> T merge() {
        return (T) JpaOperations.merge(this);
    }
```

That will enable to use it as: 

```java
    @PUT
    @Transactional
    public void update(Foo foo) {
        Foo updated = foo.merge(); //instead of: PanacheEntity updated = foo.merge();
        foo.flush();
    }
```

I think it's safe to cast the `merge()` return to `T` and therefore include a `@SuppressWarnings("unchecked")`. If we try to generify the `JpaOperations` method, we go back to the same warning again.

**Comment by fdai2003** [comment]
This will work for us to.
It requires that the developer of the update method need to know if persist or merge must be used. But in this case it is clear to call merge, because the entity is detached and modified by a user.

**Comment by emmanuelbernard** [comment]
Right. I did not bother with the generic work. But your solution is better.

**Comment by FroMage** [comment]
OK, my intention got lost in translation. Here's what I'm proposing:

```java
@Entity 
public class Foo extends PanacheEntity {
    @Version public long version;
    public String title;
    public String author;
}

// proposal
@PUT
public updateTitle(Foo foo) {
    Foo existing = Foo.findById(foo.id);
    existing.title = foo.title;
    // this throws if the versions differ if the entity is persistent
    existing.version = foo.version;
}

// this is if we add merge()
@PUT
public updateTitle(Foo foo) {
    // this would throw
    foo.merge();
}
```

What I'm saying is that we can make `setVersion` throw if the entity is attached and the version differs, which achieves the same semantics as `merge`, and does not force us to add `merge` for _this_ use-case.

This way, we can discuss adding `merge` for other use-cases, but not version checking.

> [from @Sanne] hum actually throwing an exception might not be straight forward as there's grey areas when it comes to associations: what if some of the objects in the model already exist in the session, and others don't? Is it ok to fail only after having loaded some of these?

Now, this is somewhat contingent on what `merge` does for associations. If it does version checking even for merged relations, then I doubt we can do that.

**Comment by emmanuelbernard** [comment]
> OK, my intention got lost in translation. Here's what I'm proposing:
> 
> ```java
> @Entity 
> public class Foo extends PanacheEntity {
>     @Version public long version;
>     public String title;
>     public String author;
> }
> 
> // proposal
> @PUT
> public updateTitle(Foo foo) {
>     Foo existing = Foo.findById(foo.id);
>     existing.title = foo.title;
>     // this throws if the versions differ if the entity is persistent
>     existing.version = foo.version;
> }
> 
> // this is if we add merge()
> @PUT
> public updateTitle(Foo foo) {
>     // this would throw
>     foo.merge();
> }
> ```
> 
> What I'm saying is that we can make `setVersion` throw if the entity is attached and the version differs, which achieves the same semantics as `merge`, and does not force us to add `merge` for _this_ use-case.

The problem with this model is that I'm sure someone will do something with intermediary version swaps and you will get an early exception before an effective merge or flush operation. I think I like merge a bit better because of that. Also closer to JPA.

**Comment by emmanuelbernard** [comment]
I had a conversation with @Sanne and conceptually, we feel that there is a missing operation in Hibernate ORM itself to do the kind of things you want to do - checking a detached entity state and do (partial or complete) update on an already attached entity.
So I'd recommend you use the  `entityManager().merge()` for now and we need to list this problem as one of the several to solve in Panache (and in this case Hibernate ORM more likely).

**Comment by manoelcampos** [comment]
Meanwhile, the code below is a workaround for those facing the same issue:

```java
    public static void update(SomeEntity entity) {
        final EntityManager em = JpaOperations.getEntityManager();
        SomeEntity updated = em.merge(entity);
        updated.flush();
    }
```

**Comment by nicmarti** [comment]
Just out of curiosity: any update related to this merge issue? We're facing the same issue

**Comment by emmanuelbernard** [comment]
We have not yet had time to focus on it unfortunately.

**Comment by Sanne** [comment]
Sorry I got distracted. I'll try to have this resolved - at least to not ignore the annotation - for 1.7

**Comment by guoxunbo** [comment]
mark. same problem with this. @verision is not available in `merge()` method.
I am had tried with this `merge` mthod. but it does not throw `OptimisticLockException`.
<img width="855" alt="image" src="https://user-images.githubusercontent.com/14329149/168210536-2d61f7af-f534-4b8b-a14e-acce06b9e704.png">

**Comment by guoxunbo** [comment]
any update related to this merge issue? We're facing the same issue？

**Comment by manoelcampos** [comment]
@guoxunbo while an official fix is not available, have you checked the workaround I've shown above?

**Comment by manoelcampos** [comment]
By the way, if you use the [REST Data Panache extension](https://quarkus.io/guides/rest-data-panache), the lock works out-of-the-box by just creating an empty interface (which the implementation is automatically provided by the lib):

```java
public interface CustomerResource extends PanacheEntityResource<Customer, Long> {
}
```

**Comment by guoxunbo** [comment]
@manoelcampos 
I had checked your workaround. I had also use the entitymanager.merge(entity) to solve this problem temporary.
i don't use the rest data panache extension.Is the empty interface  need to create by everyEntity? Is there has some global solution to solvethis?

**Comment by manoelcampos** [comment]
If you use REST Data Panache, you need to create that interface for each entity you want to apply the version lock.
Some other workarounds were discussed here.

**Comment by guoxunbo** [comment]
@manoelcampos 
oh. thanks. But my have many entity so if every entity create an empty interface like this.  It's have some problem to push this...
@Sanne @FroMage Does there have some plan to solve this problem.

**Comment by guoxunbo** [comment]
@manoelcampos 
hello. I had saw your code in https://github.com/manoelcampos/vendas-rest-quarkus/commit/7b8af7f74401a4c1e46d092065712b619399a074
I have a question ? is there abstractResource just ok or every entity should write a resource to extends abstractResource?
I don't know you abstractResource mean? Is there anywhere used it?

**Comment by manoelcampos** [comment]
Oh, that code is in Portuguese because it's used by my students.
But you can translate the code using Google Chrome.
The AbstractResource is an alternative to REST Data Panache (if you just want to create your own implementations for your REST controllers/resources). It provides default implementations for basic CRUD operations.
This project shows 3 different ways to create a REST controller/resource:

- from scratch (CategoriaResource), which causes lots of duplicated code for each controller/resource you implement;
- using REST Data Panache (ClienteResource) which is the easiest and fastest way;
- and using my own generic interface AbstractResource. That interface should be implemented by each REST controller/resource you want, such as MarcaResource. This is the way that provides more freedom to implement custom methods.

We can continue this discussion on my repo.

**Comment by gbourant** [comment]
any update regarding this issue?

**Comment by HallefBruno** [comment]
Estou possando pelo mesmo problema

**Comment by FroMage** [comment]
Wow, this is an old issue.

Let's try to sum it up.

- Many applications send updates of their entities to rest endpoints. The endpoint tries to persist those updates to the database.
- One way of dealing with concurrency conflicts is using `@Version`, so that we can compare the version of the updated entity with the version of the entity currently in the DB.
- Some endpoints rely on the entity being deserialisable directly from JSON/XML and so they take it as a rest endpoint parameter, which leads to the entity being "detached", and in need of being either reattached, or having its updated state transferred into an "attached" entity
- Manually finding the attached entity from the DB, and copying the udated fields to it (so that they will be persisted) leads to a complete bypass of the version checks
- Only the `merge` operation, which turns a "detached" entity into an "attached" entity achieves those version checks

I suppose I can infer the following too, correct me if I'm wrong:

- Any attempt to update an entity with `@Version` that does not use `merge` will render the version useless, and is probably a bogus update.

I seriously dislike deserialising entities directly from JSON/XML, because in most cases it's not realistic, because most rest endpoints do not allow for every entity field to be updated, or requires security permissions for different fields, or even worse: because most entity relations will simply be missing from the JSON/XML representation, due to lack of references (by ID) in those formats. Not to mention the security risks of allowing entire entities being posted via JSON/XML, possibly exposing internal fields (famous CVE source).

So in most cases, we advise people to use alternate "data transfer object" representations for transfering the entity updates to the rest endpoint, and then it's up to the endpoint to transfer those updates to JPA updates. Usually I'd have done it like this:

```java
@PUT
public void updateTitle(MyEntityDTO dto){
 MyEntity entity = MyEntity.findById(dt.id);
 // only update the title, not the other fields of the entity
 entity.title = dto.title;
}
```

But this bypasses the version checks. So people have to write this, I suppose:

```java
@PUT
public void updateTitle(MyEntityDTO dto){
 MyEntity entity = new MyEntity();
 // add the id so we can find it back in merge
 entity.id = dto.id;
 // only update the title, not the other fields of the entity
 entity.title = dto.title;
 // have to pass the version too, I'm sure
 entity.version = dto.version;
 // Apply the title change onto the attached version of this entity, found by ID, checking the version
 Panache.getEntityManager().merge(entity);
}
```

This is not ideal, and cumbersome. Also, I'm not entirely sure how `merge` figures out that only the `title` attribute was changed, and not the others (I did not see the spec mention that `merge` was a partial state copy in https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2-m2#a1983 so perhaps it's actually not the operation we're looking for).

Soooo… I'm still not convinced `merge` is what we're after for transfering updates from the rest endpoint to attached entities, with or without `@Version` attributes.

Feel free to correct me, @gavinking, @Sanne or @emmanuelbernard .

**Comment by alex123asd123asd** [comment]
Hello.
It is very interesting, but `@Version` works for me.
I'm using optimistic locking.

**Comment by manoelcampos** [comment]
IMHO, using a DTO should be a developer choice.
But the point in that Rest Data Panache just works. I suppose the merge operation is used there.

> By the way, if you use the [REST Data Panache extension](https://quarkus.io/guides/rest-data-panache), the lock works out-of-the-box by just creating an empty interface (which the implementation is automatically provided by the lib):
> 
> ```java
> public interface CustomerResource extends PanacheEntityResource<Customer, Long> {
> }
> ```

**Comment by FroMage** [comment]
> @FroMage IMHO, using a DTO should be a developer choice.

Sure. And there are cases where one is not needed. But there are also many cases where it's very dangerous and unrealistic to not use a DTO.

> But the point in that Rest Data Panache just works. I suppose the merge operation is used there.

Yes, it does. In general `merge()` will check optimistic locking, and is the only operation that does. This issue, confusingly, is not really about this, but about the fact that without `merge` there's no way to check optimistic locking, and because there are problems with `merge`, as I've noted, we're lacking a solution without `merge`.

**Comment by FroMage** [comment]
So, I suppose the real problem is a purely ORM problem and we should open an issue there to discuss.

**Comment by geoand** [comment]
cc @yrodiere

**Comment by manoelcampos** [comment]
@apetrichkovich you don't seem to be using Panache.
Using Hibernate directly just works, as I've shown [here](https://github.com/manoelcampos/hibernate-concurrency/tree/master/1-database-concurrency-plain-hibernate).

> Hello. It is very interesting, but `@Version` works for me. I'm using optimistic locking.

**Comment by yrodiere** [comment]
Hey.

@manoelcampos, from what I can see your Panache reproducer manually alters the `@Version` property: https://github.com/manoelcampos/hibernate-concurrency/blob/master/2-database-concurrency-panache/src/main/java/com/manoelcampos/server/model/Cliente.java#L27

This is explicitly forbidden and can lead to issues, probably including the one you're noticing. Quoting [the docs](https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#locking-optimistic-version-number):

> Your application is forbidden from altering the version number set by Hibernate.

The docs also explain what you're supposed to do instead:

> To artificially increase the version number, see the documentation for properties LockModeType.OPTIMISTIC_FORCE_INCREMENT or LockModeType.PESSIMISTIC_FORCE_INCREMENT in the Hibernate Entity Manager reference documentation

... though in your case, I suppose it doesn't help. From what I understand you store the entity data in a browser, including the version, and upon sending a request to your backend, you want to check that browser-stored version number against the one in the DB, to be sure nobody changed the entity in the meantime.

In order to do that, yes, you could:

1. Use `merge` -- though @FroMage will need to confirm whether that's possible using just Panache or not.
2. Manually compare the version of the entity you load from the DB -- without updating it! If the version is different, a change occurred and you want to fail.

I suppose we could imagine exposing something like `2.` directly in Panache or `Session`; something like `session.load(MyEntity.class, someId, someVersion)` that would throw an exception if the entity has a different version in the DB.

However:

* It's mostly syntactic sugar at this point.
* I'm not entirely sure optimistic locking was supposed to be used beyond transaction boundaries. Yours might be an entirely different use case than what ORM devs are trying to address. It looks like some kind of "business-level optimistic locking" to me. Maybe @sebersole could confirm/deny?

**Comment by manoelcampos** [comment]
> @manoelcampos, from what I can see your Panache reproducer manually alters the `@Version` property: https://github.com/manoelcampos/hibernate-concurrency/blob/master/2-database-concurrency-panache/src/main/java/com/manoelcampos/server/model/Cliente.java#L27
> 
> This is explicitly forbidden and can lead to issues, probably including the one you're noticing. Quoting [the docs](https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#locking-optimistic-version-number):

That was just one of the many tries to fix the issue. The optimistic lock exception is not thrown anyway, changing the version attribute or not.

> ... though in your case, I suppose it doesn't help. From what I understand you store the entity data in a browser, including the version, and upon sending a request to your backend, you want to check that browser-stored version number against the one in the DB, to be sure nobody changed the entity in the meantime.

Exactly

> 1. Use `merge` -- though @FroMage will need to confirm whether that's possible using just Panache or not.

But that is the point. Panache doesn't expose the merge method in JpaOperations class and it was said here the merge operation has its own issues.

> 2. Manually compare the version of the entity you load from the DB -- without updating it! If the version is different, a change occurred and you want to fail.

Yeah, we have discussed some approaches here, I even mentioned one I used. But anyone using Hibernate Optimistic Lock expect it to work with Panache too.

**Comment by FroMage** [comment]
> Use merge -- though @FroMage will need to confirm whether that's possible using just Panache or not.

It is possible, but its behaviour and semantics are problematic in many cases.

> I suppose we could imagine exposing something like 2. directly in Panache or Session; something like session.load(MyEntity.class, someId, someVersion) that would throw an exception if the entity has a different version in the DB.

This would be a valid approach. As I've noted, versions appear to only ever work when using `merge`, which is not necessarily intuitive for people. That would be a useful operation to add to ORM, but we can start by adding it in Panache.

On the other hand, this `load` will only throw an error if the version does not match at read time, remains to be seen what happens if someone commits a write operation that bumps this version by the time our current transaction commits. I hope it throws again at this time, but I haven't checked.

**Comment by yrodiere** [comment]
> Yeah, we have discussed some approaches here, I even mentioned one I used. But anyone using Hibernate Optimistic Lock expect it to work with Panache too.

My point is that this feature is simply not there in Panache. It's not that it "doesn't work", it's just not there.

Which doesn't solve anything for you, I know. But it means we're talking about exposing a feature here, not fixing incorrect behavior.

> > Use merge -- though @FroMage will need to confirm whether that's possible using just Panache or not.
> 
> It is possible, but its behaviour and semantics are problematic in many cases.

Can't argue with that. `merge` is a great way to shoot yourself in the foot.

> That would be a useful operation to add to ORM, but we can start by adding it in Panache.

+1 for a feature request -- anyone? :) https://hibernate.atlassian.net/browse/HHH

> As I've noted, versions appear to only ever work when using merge, which is not necessarily intuitive for people.

Keep in mind changing the version manually in an attached entity is forbidden. So the _only_ way to have Hibernate ORM witness a version conflict right now is, in fact, by using `merge`.

> On the other hand, this `load` will only throw an error if the version does not match at read time, remains to be seen what happens if someone commits a write operation that bumps this version by the time our current transaction commits. I hope it throws again at this time, but I haven't checked.

I'd expect a write to fail if the version changed between the read and the write, yes.

**Comment by jacksoncastro** [comment]
## Problem

I also had problems with `Optimistic Lock` with Panache in Quarkus.

My environment is as follows:

- Quarkus `3.11.1`
- hibernate-core `6.5.2`

> I also tested with the latest version `3.21.3` of quarkus

Here is my example entity:

```java
@Entity
public class MyEntity extends PanacheEntityBase {

    @Id
    @GeneratedValue
    public Long id;

    public String field;

    @Version
    public long version;
}
```

When doing what most documentation suggests, no exception is thrown for `Optimistic Locking`:

```java
@Path("/entity")
public class MyEntityResource {
    @Transactional
    @PUT
    @Consumes(MediaType.APPLICATION_JSON)
    public Response update(MyEntityDTO dto) {
        MyEntity myEntity = (MyEntity) MyEntity.findByIdOptional(id).orElseThrow(NotFoundException::new);
        mapper.toEntity(dto, myEntity);
        myEntity.persist();
    }
}
```

This does not work because the `MyEntity` entity is **managed** by Hibernate.

During my debugging I saw that at the moment the transaction is going to be committed the following section will be called in `org.hibernate.event.internal.DefaultMergeEventListener.merge(MergeEvent, MergeContext, Object)`:

```java
switch ( entityState ) {
    case DETACHED:
        entityIsDetached( event, copiedId, originalId, copiedAlready );
        break;
    case TRANSIENT:
        entityIsTransient( event, copiedId != null ? copiedId : originalId, copiedAlready );
        break;
    case PERSISTENT:
        entityIsPersistent( event, copiedAlready );
        break;
    default: //DELETED
```

If the entity `entityState` is `PERSISTENT` it enters `entityIsPersistent` which in turn does not validate if the version has been changed. However, if it falls into the `DETACHED` flow, `entityIsDetached` will call `org.hibernate.event.internal.DefaultMergeEventListener.targetEntity(MergeEvent, Object, EntityPersister, Object, Object)`, finally validating the versioning.

## Solution

Two things are important for it to work:

- The entity cannot be managed. In other words, do not fetch from the database and replace the values, create an object manually;

- Only the `merge` method of the `EntityManager` makes the entity enter `DETACHED` correctly.

```java
@Path("/entity")
public class MyEntityResource {

    @Inject
    private EntityManager entityManager;

    @Transactional
    @PUT
    @Consumes(MediaType.APPLICATION_JSON)
    public Response update(MyEntityDTO dto) {
        if (!repository.existsById(dto.getId())) {
            throw new NotFoundException();
        }
        MyEntity myEntity = mapper.toEntity(dto)
        entityManager.merge(myEntity);
    }
}
```

I don't know if this is a bug or expected behavior, but this is how it worked for me.

As a bonus tip, create an abstraction for your entities to extend with the `update` method:

```java
public abstract class EntityGenericAbstract<T> extends PanacheEntityBase {
    public void update() {
        getEntityManager().merge(this);
    }
}
```

Then just call it directly:

```java
myEntity.update();
```

**Comment by FroMage** [comment]
This is useful info for anyone insisting to use `merge`, thanks :)

**Comment by geoand** [comment]
> My point is that this feature is simply not there in Panache. It's not that it "doesn't work", it's just not there.

> Which doesn't solve anything for you, I know. But it means we're talking about exposing a feature here, not fixing incorrect behavior.

Based on this I am changing this from `bug` to `enhancement`
